### DISCUSSION TITLE
Const strings are not so const after all

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#97] [General]
# Const strings are not so const after all

We know the const keyword with strings - but are they really constant after all? Or can we use some tricks to modify them? Let's see.

https://steven-giesel.com/blogPost/544c33da-be2e-4d96-99c9-9313f48548bd/const-strings-are-not-so-const-after-all

**Comment by toupswork** [comment]
I'm pretty sure that you may not always get an access violation if you overrun it. If you're writing production code, that's what you hope happens. It depends where that memory is. Based on that error, it looks like it overran on into a virtual memory page that was marked as execution only. That's called DEP (Data Execution Protection) that Microsoft added to Windows a long time back during the buffer overrun days. But if the location is on a page marked to hold data, you might not. Also if the location wasn't assigned to a page file, then you also get an error.

**Reply by linkdotnet** [reply]
That might be a correct assumption. 
in theory the runtime knows if you go out of bounds with your access to the memory location so I for sure hope that it does something rather than letting it happen silently.

Do you know how Linux handles that?
My given experience here is with MacOS.

**Reply by toupswork** [reply]
No I don't. I've always been a Windows person. But I am pretty sure they work the same. What I remember is that you call malloc and the OS works with the MMU to give you pages that back the virtual addresses that malloc returns. The virtual addresses it gives you have to be above a certain point. It won't give addresses in the lower range, partly because that's often where bad pointers will point and partly for historical reasons. I think Microsoft worked with Intel to get them to add instructions for DEP, so that should be cross-platform..  That was around the year 2004 when there was constant buffer overrun attacks coming out. The virtual memory pages have attributes assigned to them by the processor that indicate if they're read/write/execute, and the MMU detects when this is broken or when you're accessing an address not backed by a page and raises a page fault error I believe.

**Reply by toupswork** [reply]
Interesting side note, once the buffer overrun days were over, some of the new attacks had to do with using branch prediction and CPU cache. They have the ability to “read” memory belonging to other processes not by actually reading it but by causing the processor to perform a branch prediction in a kernel-mode switch to cause the processor to readahead in kernel memory and load it into the CPU cache. Then through clever trickery, they can use a high-resolution timer to see how long it takes for an area of memory to be read, and clocking how quickly the memory comes back they can in-effect infer what memory exists in kernel space. It's super crazy..

**Reply by linkdotnet** [reply]
Didn’t they even went so far to use such exploits with JavaScript aka the RowHammer library? Sure it is not the same as „misusing“ branch prediction- but goes in a similar way.

It is really crazy how you can extract information with timing based attacks.

**Reply by toupswork** [reply]
Yep you're right. So Chrome and other browsers had to lower the timer resolution. I know, it's wild.

**Reply by toupswork** [reply]
Then they also did some other attack where rather than inserting rogue code, they take the code that's already there and then use jumps and returns to string together that existing code into an exploit..

**Reply by linkdotnet** [reply]
It is wild - and unfortunately far behind my comprehension!
It is crazy to see how one can come up with such an exploit

**Reply by toupswork** [reply]
No it's not. Believe me.. I thought the same thing of myself. But I read the whitepaper on it, and I thought it was going to be so overly academic and incredibly dry. I was shocked at how easy it was to comprehend it. I thought I would just spend a couple of minutes glancing through it and then my eyes would glaze over and I'd put it down.. But no it wasn't that hard to understand.

**Reply by linkdotnet** [reply]
Well then I will put this onto my reading list.

Maybe I can combine this with another interest of mine GPGPU (Cuda and OpenCL). It seems you can do the same in the RAM of a graphics card as well.

**Reply by toupswork** [reply]
That's sounds need. When .NET started adding SIMD intrinsics support, I became a bit fascinated with vector-based programming. I studied linear algebra a while back, so I wasn't completely lost. I wrote a custom hash function that uses vectors to create a 64bit hash code that I covert to ulong. It was really cool. I didn't want to use the default hash algorithm in .NET, called “Martin“, because it's no longer deterministic, and I want the same hash to be generated on multiple nodes. Plus you probably know about the memory layout for a class in .NET and how the hashcode is placed on the header and it only has 26bits of entropy it can store there. When our API makes calls to aggregate data from other APIs, I am caching the responses. I created a GetOrAddGet method, and I needed a locking mechanism to prevent the avalanche effect if multiple cache requests come in at the same time for the same thing and start stacking up. I want only the first thread to be allowed time to check the cache and fetch the response and store it in redis if it's a cache miss. So what I did was take the URI of the API, the HTTP Verb Name, and depending on if the cache can be shared across users or not, I take the authorization header from incoming request, concatenate it to a string, then generate the hashcode from it.  Then I created a FrozenSet of 21 SemaphoreSlim objects, and I take the ulong hash for the request and MOD over 21 to have it grab a semaphoreslim from the pool.. So the same requests will get the same semaphoreslim from the pool. Then the thread can add the response to cache and release the semaphore. It works pretty well.. But vectorization and matrices are very interesting. What do you plan on using it for? AI? Gaming?

**Reply by linkdotnet** [reply]
I guess (in case of strings) they do this for security reasons.

you choose this approach so you don’t have to create a new semaphore per HTTP Call, right!? 
as far as I remember the team wanted to introduce a stable hashes for strings that stays the same over multiple runs. But as you already have the logic - roll with it.

it started in my university time where we used CUDA to identify bees in realtime on a 2D image (nothing fancy nor image recognition - the bees hat literally a qr code painted on their back and they only moved in a 2d surface).

If you don’t have massive data SSE is a really life start. I used SSE for LINQ like operations: https://github.com/linkdotnet/LinqSIMDExtensions

then there is also ILGPU which allows GPGPU in .net. Internally they translate it to CUDA or OpenCL. 
I don’t have a use case at hand - it is just a very interesting topic for me.

**Reply by toupswork** [reply]
Right, that's what I read. They changed it to be non-deterministic for security reasons. Think about it this way, say that the Cache Provider has a SemaphoreSlim field in it. That becomes a choke point that will reduce throughput. The lock has to be held to (1) check for an entry in cache, (2) if it's a cache miss it needs to go out and make the actual request to the destination API, (3) add that response to the cache and then release the semaphore. That a long time to wait especially if there is heavy traffic. With heavy traffic every request is most likely going to have to wait on that semaphore to be released. But if I have a pool of semaphores and I ensure that the same requests grab the same semaphore from the pool, now I have just increased throughput and reduced contention. The goal is to ensure that only the same requests from the same resource have to wait for the lock to release. So I needed a high entropy hash to reduce the standard deviation as low as I can get it so that there is no bias toward any particular number and it's evenly spread. My hash has a lower standard deviation over a pretty large sample size than the Martin hash, plus the all the nodes generate the same hash for the same request.

**Reply by toupswork** [reply]
Yeah I saw that project. I thought it was cool. I was looking for a chance to use it, so the hash was my first opportunity.

**Reply by toupswork** [reply]
Wow that sounds really cool. You must have gone to grad school for CS?

**Reply by linkdotnet** [reply]
Thanks for the explanation - I guess I got it.
I am interested why you need the Hashcode rather than using your constructed string as key?

The bee project was part of my bachelor thesis and I assisted a PhD. So much of the ground work was somewhat done, but they needed a somewhat realtime analysis of the pictures.
Basically some 12MP images per second and they needed the position of every bee in each picture (under 1 second before the next image comes in).
That is somewhat 12 years ago where 12 MP wasn’t your goto resolution for pictures.

**Reply by toupswork** [reply]
Because I don't want the concatenated string that contains the authorization header and the URI info in the Redis database for the key because it's not only a long string but it contains sensitive info. I want a short Base64 string for the key. So the hash bytes are converted to both a base64 string for the key and a `ulong` for picking a lock from the pool. 

So basically you were using a neural network to identify the bees in the image?

**Reply by linkdotnet** [reply]
Fair point about sensitive data. I was just wondering.

No we did not use any fancy ML stuff back in those days.
We basically used CUDA to make the Oma he grayscale and did edge detection on the picture so it is way more convenient to find the we codes.
Then we just counted the pixels until we found the pattern we were looking for. Basically something like: „Hey I need to find 1x white pixels then 3x black ones, theb 1x white again and so on“
So the ratio between those were important - if you found that, you found the QR Code and then just grab the important data out of it.
Very naive approach.
